gpflow/models/vgp.py

```diff
@@ -70,12 +70,67 @@ def __init__(
 
         self.data = data_input_to_tensor(data)
         X_data, Y_data = self.data
-        num_data = X_data.shape[0]
-        self.num_data = num_data
 
-        self.q_mu = Parameter(np.zeros((num_data, self.num_latent_gps)))
-        q_sqrt = np.array([np.eye(num_data) for _ in range(self.num_latent_gps)])
-        self.q_sqrt = Parameter(q_sqrt, transform=triangular())
+        # If the input data is variable, then the user probably intends to mutate it, and then we'll
+        # need a copy to keep consistency between it and the `self.q_*` variables.
+        def copy_if_variable(t: TensorType) -> TensorType:
+            if is_variable(t):
+                return Parameter(t, shape=t.shape, trainable=False)
+            return t
+
+        self.internal_data = (
+            copy_if_variable(X_data),
+            copy_if_variable(Y_data),
+        )
+
+        static_num_data = X_data.shape[0]
+        if static_num_data is None:
+            self.num_data = Parameter(
+                tf.shape(X_data)[0], shape=[], dtype=tf.int32, trainable=False
+            )
+            q_sqrt_transformed_shape = (self.num_latent_gps, None)
+        else:
+            self.num_data = tf.constant(static_num_data)
+            q_sqrt_transformed_shape = (self.num_latent_gps, triangular_size(self.num_data))
+        # Many functions below don't like `Parameter`s:
+        dynamic_num_data = tf.convert_to_tensor(self.num_data)
+
+        self.q_mu = Parameter(
+            tf.zeros((dynamic_num_data, self.num_latent_gps)),
+            shape=(static_num_data, num_latent_gps),
+        )
+        q_sqrt = tf.eye(dynamic_num_data, batch_shape=[self.num_latent_gps])
+        self.q_sqrt = Parameter(
+            q_sqrt,
+            transform=triangular(),
+            pretransformed_shape=(num_latent_gps, static_num_data, static_num_data),
+            transformed_shape=q_sqrt_transformed_shape,
+        )
+
+    def after_data_changed(self) -> None:
+        new_X_data, new_Y_data = self.data
+        new_num_data = tf.shape(new_X_data)[0]
+        f_mu, f_cov = self.predict_f(new_X_data, full_cov=True)  # [N, L], [L, N, N]
+
+        # This model is hard-coded to use the whitened representation, i.e.  q_mu and q_sqrt
+        # parametrize q(v), and u = f(X) = L v, where L = cholesky(K(X, X)) Hence we need to
+        # back-transform from f_mu and f_cov to obtain the updated new_q_mu and new_q_sqrt:
+        Knn = self.kernel(new_X_data, full_cov=True)  # [N, N]
```

Comment:
This is going to significantly simplify the trieste `VariationalGaussianProcess` wrapper class :-)

Reply:
> This is going to significantly simplify the trieste `VariationalGaussianProcess` wrapper class :-)

I know. Did you seem my proof-of-concept PR? :-)